### ISSUE TITLE
Problems with MoveRefactoring + Package name quickfix

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-xtend#696]
1. create a class in package a
2. move it to package b
3. use the fix package name quickfix
4. generated java file and trace files stay in the old packaged

Expected: Java code and trace files in old packages are gone

**Comment by Hanksha** [comment]
Is it normal that the package declaration is not refactored? I always have to use the quickfix for it, it's a bit annoying when you move a lot of file.

**Comment by cdietrich** [comment]
Yes there is no move refactoring implemented right now

**Comment by Hanksha** [comment]
If I can be of any help I can give it a try. Do I need to implement a `org.eclipse.ltk.core.refactoring.participants.MoveRefactoring` for xtend files?

**Comment by cdietrich** [comment]
I assume so. But I don’t know how proper this will work and why it was not implemented in the first place. @JanKoehnlein do you have any idea?

**Comment by JanKoehnlein** [comment]
I implemented the refactorings in Xtend quite a while ago. I remember there were good reasons to skip move refactoring, but not exactly which ones :-( Must have been something like dragging and dropping files in the tree did not trigger any kind of refactoring or at least not the correct one such that it was too hard too hook into this. Or that JDT only considers a source package relevant when there are Java resources in it. I am pretty sure there were problems on these early stages already, and we expected them to multiply the more we move forward with the implementation.

All in all, the refactorings in Xtend are _not_ something I am particularly proud of: They are pretty hacky, because the base API from LTK does not really match the usecases, and it has to be combined with JDT that does everything very differently behind these APIs and it does not really allow to hook into the right places. We even had to store state in static variables, such that they can be picked up later, as there was no communication channel between the various steps. I'd rather not touch any of this. 

As Xtend has a lot of static analysis, it is usually feasible (yet annoying) to clean up after a file move manually.

**Comment by JanKoehnlein** [comment]
Sounds easy, doesn't it? Feel free to try :-) The crucial part is the last sentence. There is no inverse compiler that reproduces exactly the original Xtend code from generated Java, let alone if you modified the latter. I don't think its feasible covering all usecases (e.g. all Java generation target versions of Xtend) in a stable manner within reasonable resources to implement it. 

Refactoring is never only textual query/replace, as scopes could change as a side effect and make different things referable under the same name. If you add too much magic, your users will be very surprised about the strange decisions you made that break their corner cases.

Did other JVM languages (e.g. Scala) ever go so far and integrate into rename refactoring of Eclipse? I doubt so. At least they didn't when I wrote the Xtend refactoring. They might have had their reasons, too.

**Comment by kthoms** [comment]
> Why not delegate the move refactoring to JDT by tricking it into moving the generated java file (in xtend-gen). You could then analyze the changes it returned and adapt them for the xtend resources.

Sounds complicated, I doubt that this is a way to go. I'd better start off analyzing the existing refactorings.

**Comment by Hanksha** [comment]
> Sounds easy, doesn't it? 

Won't be easy for sure. I've done my fair share of hacking around with JDT and the Groovy Eclipse plugin and it was a pain in the ass every time.

I am not proposing to analyze the modified generated java source but the encapsulated changes. I have implemented move refactoring for several visual languages and I encapsulated the changes to be applied on the models in `org.eclipse.ltk.core.refactoring.Change` objects. If JDT took the same approach maybe there is some stuff that could be reused.

I can understand integrating Xtend refactoring with other projects like JDT and PDE etc. is not a priority but even move refactoring a Xtend file within the same Xtend project doesn't refactor its package name (at least for me) which is IMO a minimum.

**Comment by kthoms** [comment]
Would be awesome if you could contribute such a refactoring!

**Comment by Hanksha** [comment]
@kthoms Yes I'll give it a try when I find some free time. I'll have to get familiar with Xtend codebase first.

**Comment by miklossy** [comment]
See also [[xtend][refactoring] implement move refactoring for types](https://bugs.eclipse.org/bugs/show_bug.cgi?id=421032)
